services/notifications/pkg/service/job.go

```diff
@@ -0,0 +1,161 @@
+	if err != nil {
+		logger.Error().Err(err).Msg("could not get list of keys")
+		return
+	}
+
+	ctx := context.Background()
+
+	jobs := make(chan string, 10)
+	go func() {
+		for _, key := range keys {
+			jobs <- key
+		}
+		close(jobs)
+	}()
+
+	for job := range jobs {
+		go s.createGroupedMail(ctx, logger, job)
+	}
+}
+
+func (s eventsNotifier) createGroupedMail(ctx context.Context, logger log.Logger, key string) {
+	userEvents, err := s.userEventStore.pop(ctx, key)
+	if err != nil {
+		logger.Error().Err(err).Str("key", key).Msg("could not pop user events")
+		return
+	}
+
+	var mts []email.MessageTemplate
+	var mtsVars []map[string]string
+	locale := l10n.MustGetUserLocale(ctx, userEvents.User.GetId().GetOpaqueId(), "", s.valueService)
+
+	for _, e := range userEvents.Events {
+		switch te := s.unwrapEvent(logger, e).(type) {
+		case events.SpaceShared:
+			logger := log.Logger{Logger: logger.With().
+				Str("event", "SpaceShared").
+				Str("eventId", te.ID.OpaqueId).
+				Logger()}
+
+			executant, spaceName, shareLink, _, err := s.prepareSpaceShared(logger, te)
+			if err != nil {
+				logger.Error().Err(err).Msg("could not prepare vars for grouped email")
+				continue
+			}
+
+			mts = append(mts, email.SharedSpace)
+			mtsVars = append(mtsVars, map[string]string{
+				"SpaceSharer": executant.GetDisplayName(),
+				"SpaceName":   spaceName,
+				"ShareLink":   shareLink,
+			})
+		case events.SpaceUnshared:
+			logger := log.Logger{Logger: logger.With().
+				Str("event", "SpaceUnshared").
+				Str("eventId", te.ID.OpaqueId).
+				Logger()}
+
+			executant, spaceName, shareLink, _, err := s.prepareSpaceUnshared(logger, te)
+			if err != nil {
+				logger.Error().Err(err).Msg("could not prepare vars for grouped email")
+				continue
+			}
+			mts = append(mts, email.UnsharedSpace)
+			mtsVars = append(mtsVars, map[string]string{
+				"SpaceSharer": executant.GetDisplayName(),
+				"SpaceName":   spaceName,
+				"ShareLink":   shareLink,
+			})
+		case events.SpaceMembershipExpired:
+			mts = append(mts, email.MembershipExpired)
+			mtsVars = append(mtsVars, map[string]string{
+				"SpaceName": te.SpaceName,
+				"ExpiredAt": te.ExpiredAt.Format("2006-01-02 15:04:05"),
+			})
+		case events.ShareCreated:
+			logger := log.Logger{Logger: logger.With().
+				Str("event", "ShareCreated").
+				Str("eventId", te.ItemID.OpaqueId).
+				Logger()}
+
+			owner, shareFolder, shareLink, _, err := s.prepareShareCreated(logger, te)
+			if err != nil {
+				logger.Error().Err(err).Msg("could not prepare vars for grouped email")
+				continue
+			}
+			mts = append(mts, email.ShareCreated)
+			mtsVars = append(mtsVars, map[string]string{
+				"ShareSharer": owner.GetDisplayName(),
+				"ShareFolder": shareFolder,
+				"ShareLink":   shareLink,
+			})
+		case events.ShareExpired:
+			logger := log.Logger{Logger: logger.With().
+				Str("event", "ShareCreated").
+				Str("eventId", te.ItemID.OpaqueId).
+				Logger()}
```

Comment:
Curious: Why do we need to wrap this in a `log.Logger{}` again? Wouldn't it be sufficient to use 
```suggestion
			logger := logger.With().
				Str("event", "ShareCreated").
				Str("eventId", te.ItemID.OpaqueId).
				Logger()
```

Reply:
Fixed 👍